utils/connection/server_util.go

```diff
@@ -70,11 +76,23 @@ func (c *ServerConfig) GrpcServer() (*grpc.Server, error) {
 }
 
 // Listener instantiate a [net.Listener] and updates the config port with the effective port.
-func (c *ServerConfig) Listener() (net.Listener, error) {
+// If the port is predefined, it will retry to bind to the port until successful or until the context ends.
+func (c *ServerConfig) Listener(ctx context.Context) (net.Listener, error) {
 	if c.preAllocatedListener != nil {
 		return c.preAllocatedListener, nil
 	}
-	listener, err := net.Listen(grpcProtocol, c.Endpoint.Address())
+
+	var err error
+	var listener net.Listener
+	if ctx == nil || c.Endpoint.Port == 0 {
```

Comment:
It is not intuitive that when the ctx is nil, retry till successful binding is not executed. I notice all callers to pass the ctx in the current code. Do we need this check?